core/src/main/java/com/scalar/db/transaction/common/ActiveTransactionManagedTransactionManager.java

```diff
@@ -0,0 +1,153 @@
+package com.scalar.db.transaction.common;
+
+import com.scalar.db.api.Delete;
+import com.scalar.db.api.DistributedTransaction;
+import com.scalar.db.api.Get;
+import com.scalar.db.api.Mutation;
+import com.scalar.db.api.Put;
+import com.scalar.db.api.Result;
+import com.scalar.db.api.Scan;
+import com.scalar.db.config.DatabaseConfig;
+import com.scalar.db.exception.transaction.CommitException;
+import com.scalar.db.exception.transaction.CrudException;
+import com.scalar.db.exception.transaction.RollbackException;
+import com.scalar.db.exception.transaction.TransactionException;
+import com.scalar.db.exception.transaction.UnknownTransactionStatusException;
+import com.scalar.db.util.ActiveExpiringMap;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class ActiveTransactionManagedTransactionManager
+    extends AbstractDistributedTransactionManager {
+
+  private static final long TRANSACTION_EXPIRATION_INTERVAL_MILLIS = 1000;
+
+  private static final Logger logger =
+      LoggerFactory.getLogger(ActiveTransactionManagedTransactionManager.class);
+
+  private final ActiveExpiringMap<String, DistributedTransaction> activeTransactions;
+
+  public ActiveTransactionManagedTransactionManager(DatabaseConfig config) {
+    activeTransactions =
+        new ActiveExpiringMap<>(
+            config.getActiveTransactionManagementExpirationTimeMillis(),
+            TRANSACTION_EXPIRATION_INTERVAL_MILLIS,
+            t -> {
+              logger.warn("the transaction is expired. transactionId: {}", t.getId());
+              try {
+                t.rollback();
+              } catch (RollbackException e) {
+                logger.warn("rollback failed", e);
+              }
+            });
+  }
+
+  private void addActiveTransaction(DistributedTransaction transaction)
+      throws TransactionException {
+    if (activeTransactions.putIfAbsent(transaction.getId(), transaction) != null) {
+      transaction.rollback();
+      throw new IllegalStateException(
+          "The transaction already exists. transactionId: " + transaction.getId());
+    }
+  }
+
+  private void removeActiveTransaction(String transactionId) {
```

Comment:
Same as above

Reply:
Fixed it in 5644f5d3ff7ab1883ce588b199639c8124bf8ad4. Thanks.